add forget bias option to LSTM cell

geoalgo opened the pull request.

piiswrong commented on python/mxnet/rnn/rnn_cell.py:
```diff
@@ -404,7 +409,7 @@ def __call__(self, inputs, states):
         in_gate = symbol.Activation(slice_gates[0], act_type="sigmoid",
                                     name='%si'%name)
-        forget_gate = symbol.Activation(slice_gates[1], act_type="sigmoid",
+        forget_gate = symbol.Activation(slice_gates[1] + self._forget_bias, act_type="sigmoid",
```
why do we need this? i2h and h2h already have biases

geoalgo replied:
True in theory you can learn it, but this optimization trick is often used in practice (in Tensorflow, it is initialized to 1.0 by default [for instance](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/contrib/rnn/python/ops/rnn_cell.py). An alternative would be to do this with a custom initializer but this is less convenient.

From the Jozefowicz et al. 2015 paper:
"There is an important technical detail that is rarely mentioned in discussions of the LSTM, and that is the initialization of the forget gate bias bf. Most applications of LSTMs simply initialize the LSTMs with small random weights which works well on many problems. But this initialization effectively sets the forget gate to 0.5. This introduces a vanishing gradient with a factor of 0.5 per timestep, which can cause problems whenever the long term dependencies are particularly severe (such as the problems in Hochreiter & Schmidhuber (1997) and Martens & Sutskever (2011)).
This problem is addressed by simply initializing the forget gates bf to a large value such as 1 or 2. By doing so, the forget gate will be initialized to a value that is close to 1, enabling gradient flow. This idea was present in Gers et al. (2000), but we reemphasize it since we found many practitioners to not be familiar with it."

piiswrong replied:
I see. But this makes it incompatible with FusedRNN. Let's use a custom initializer instead. It's not hard:
bias = mx.Variable('_bias', init=LSTMbias)
Also we need a way to handle this for fusedrnncell.

piiswrong replied:
Also with initializer we don't need to worry about backward compatibility since its retraining anyway. We can set the default to 1.

geoalgo replied:
Thanks for the suggestion, I'll send a CR with the initializer instead and I'll try to add to FusedRNNCell (its done in TF).
In the long run, do we plan to have identical features of LSTMCell in FusedRNNCell? We may be limited by the API of the Cudnn symbol.RNN when including other LSTM functionalities, maybe we could add another LSTM with richer set of functionalities that is not compatible with FusedRNNCell?

geoalgo replied:
By incompatible with FusedRNN, do you mean that the forget bias will be applied in LSTMCell but not in FusedRNNCell(LSTMCell)? Or is there another issue?

piiswrong replied:
yes that's what I meant.

We plan to keep compatibility where possible. Otherwise LSTMCell should be a super set of FusedRNNCell.

geoalgo replied:
Ok, I see but does using "bias = mx.Variable('_bias', init=LSTMbias)" solve it? 

So far, I only see one way to do it by relying on a forget bias name convention in FusedRNN._init_bias but it is ugly. Do you have a better idea in mind?

szha replied:
Could you instead make a special initialization function for this purpose? Since the shapes are predefined, you can just concatenate zeros and ones to do the initialization without breaking the compatibility.

piiswrong replied:
For LSTMCell you can do init=LSTMbias. For FusedRNN unfortunately I can't think of a better way.

geoalgo commented:
Sure I'll try to send an updated version with your suggestion tomorrow.

On Wed, Mar 15, 2017 at 8:43 PM, Sheng Zha <notifications@github.com> wrote:

> *@szha* commented on this pull request.
> ------------------------------
>
> In python/mxnet/rnn/rnn_cell.py
> <https://github.com/dmlc/mxnet/pull/5395#discussion_r106264777>:
>
> > @@ -404,7 +409,7 @@ def __call__(self, inputs, states):
>                                            name="%sslice"%name)
>          in_gate = symbol.Activation(slice_gates[0], act_type="sigmoid",
>                                      name='%si'%name)
> -        forget_gate = symbol.Activation(slice_gates[1], act_type="sigmoid",
> +        forget_gate = symbol.Activation(slice_gates[1] + self._forget_bias, act_type="sigmoid",
>
> Could you instead make a special initialization function for this purpose?
> Since the shapes are predefined, you can just concatenate zeros and ones to
> do the initialization without breaking the compatibility.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/dmlc/mxnet/pull/5395#discussion_r106264777>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AJXUz-HWRJHtpM1jZ1fOVCXuFUUtfaBgks5rmD9MgaJpZM4Mc5w6>
> .
>

piiswrong commented on python/mxnet/initializer.py:
```diff
@@ -431,8 +432,30 @@ def _init_weight(self, _, arr):
+
+    forget_bias: float, bias for the forget gate, Jozefowicz et al. 2015 recommends setting this to 1.0.
+    """
+    def __init__(self, num_hidden, forget_bias):
```
I think it's safe to divide by 4 and omit the num_hidden arg here since we are assuming LSTM anyway

piiswrong commented on python/mxnet/initializer.py:
```diff
@@ -467,5 +491,12 @@ def _init_weight(self, _, arr):
             desc = InitDesc(name)
-            self._init(desc, args[name])
+            # for lstm bias, we use a custom initializer which adds a bias to the forget gate
+            if self.mode == 'lstm' and name.endswith("bias"):
```
unpack will also unpack the bias, so there will be _i_bias/_f_bias, etc

geoalgo commented:
I have rewritten it with a custom initializer (thanks for the suggestion) and added it also for FusedRNN. I have added a test for both, let me know if it sounds correct to you.

szha commented on python/mxnet/initializer.py:
```diff
@@ -431,8 +432,32 @@ def _init_weight(self, _, arr):
 
 
 @register
+class LSTMCellForgetBias(Initializer):
```
Consider dropping "ForgetBias" from the name in case other LSTM param initialization want to reuse this.

piiswrong replied:
LSTMBias should be good.

szha replied:
The name "LSTMBias" means that we won't add weight initialization logics in it. Also the naming convention would be slightly different than FusedRNN(Initializer).

szha commented on python/mxnet/rnn/rnn_cell.py:
```diff
@@ -355,14 +355,20 @@ class LSTMCell(BaseRNNCell):
+
     """
-    def __init__(self, num_hidden, prefix='lstm_', params=None):
+    def __init__(self, num_hidden, prefix='lstm_', params=None, forget_bias=1.0):
```
Instead of adding new parameter to the constructors and introducing new default behavior, how about the following:
- extend RNNParams (e.g. LSTMParams)
- Add the initialization behavior to the constructor of that new class
- When using this, user invokes the constructors with an instance of the new Params class.

piiswrong replied:
There is no need for the complications yet. We can always refactor it latter if we need to over load that behavior

szha replied:
OK. I brought this up mostly from engineering perspective, as it's generally not recommended to change the default behavior, since codes that are already using this API have no way of knowing the changes underneath. New RNNParams class seems to be a natural way of encapsulating new changes while maintaining the current behavior.

That said, I'm fine either way since users would most likely welcome a better default parameter. API-wise, keep in mind that once you add a new argument, there's really no turning back, because removing the argument would break codes that started to use it. Semantic-wise, though I prefer keeping parameter initialization in the parameter class, OOP-wise I think it's still OK to say that "parameter initialization is the concern of an LSTM cell".

piiswrong commented on python/mxnet/rnn/rnn_cell.py:
```diff
@@ -355,14 +355,20 @@ class LSTMCell(BaseRNNCell):
         self._hW = self.params.get('h2h_weight')
-        self._hB = self.params.get('h2h_bias')
+        forget_bias_init = init.LSTMCellForgetBias(num_hidden=num_hidden, forget_bias=forget_bias)
+        self._iB = self.params.get('i2h_bias', init=forget_bias_init)
```
Only needed for one of iB and hB right? otherwise it will be doubled.

geoalgo replied:
indeed

piiswrong commented on python/mxnet/rnn/rnn_cell.py:
```diff
@@ -355,13 +355,19 @@ class LSTMCell(BaseRNNCell):
     params : RNNParams or None
         container for weight sharing between cells.
         created if None.
+    forget_bias : bias added to forget gate, default 0.0.
```
default 1.0

geoalgo replied:
Updated

geoalgo commented:
Is there a recommended way to run GPU tests? I push multiple times to get the test results as only GPU tests were failing but it does not seems like a great way...

piiswrong commented:
nosetests --verbose tests/python/gpu/test_operator_gpu.py

piiswrong commented on python/mxnet/initializer.py:
```diff
@@ -438,25 +483,32 @@ class FusedRNN(Initializer):
+            # for lstm bias, we use a custom initializer
+            # which adds a bias to the forget gate
+            if self._mode == 'lstm':
+                FusedLSTMBias(forget_bias=self._forget_bias)._init_weight(desc, args[name])
```
this is initializing weights with FusedLSTMBias.
we don't need a separate FusedLSTMBias. just set to 0 when you see *_f_bias

geoalgo replied:
I have removed it. 
Is anyone working on adding simple regression tests on RNN? I have only seen regression test for conv net, we could check an error metric on a simple datasets training with GRU/LSTM.

geoalgo replied:
For the tests, I meant if you dont have a gpu. I guess you have to spawn an instance then.

geoalgo commented:
Do you still see issues for this PR to be fixed?

piiswrong commented on python/mxnet/initializer.py:
```diff
@@ -438,25 +462,32 @@ class FusedRNN(Initializer):
         should be the same with arguments passed to FusedRNNCell.
     """
-    def __init__(self, init, num_hidden, num_layers, mode, bidirectional=False):
+    def __init__(self, init, num_hidden, num_layers, forget_bias, mode, bidirectional=False):
```
actually this breaks loading of previously saved symbols. forget_bias needs to be the last argument with a default value.

geoalgo replied:
should we add a test that loads saved symbols for LSTM/GRU?

piiswrong commented on python/mxnet/rnn/rnn_cell.py:
```diff
@@ -512,7 +516,7 @@ class FusedRNNCell(BaseRNNCell):
     ----------
     """
-    def __init__(self, num_hidden, num_layers=1, mode='lstm', bidirectional=False,
+    def __init__(self, num_hidden, num_layers=1, forget_bias=1.0, mode='lstm', bidirectional=False,
```
move forget_bias to after initializer

piiswrong commented:
I just notice an issue.
Some last thing and we can merge.

piiswrong commented:
yes also need to init Param after load

David Salinas <notifications@github.com>于2017年3月21日 周二上午11:42写道：

> *@geoalgo* commented on this pull request.
> ------------------------------
>
> In python/mxnet/initializer.py
> <https://github.com/dmlc/mxnet/pull/5395#discussion_r107241638>:
>
> > @@ -438,25 +462,32 @@ class FusedRNN(Initializer):
>      bidirectional : bool
>          should be the same with arguments passed to FusedRNNCell.
>      """
> -    def __init__(self, init, num_hidden, num_layers, mode, bidirectional=False):
> +    def __init__(self, init, num_hidden, num_layers, forget_bias, mode, bidirectional=False):
>
> should we add a test that loads saved symbols for LSTM/GRU?
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/dmlc/mxnet/pull/5395#discussion_r107241638>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAiudEAMwyyxqDePP5oBU0qBacgVueCtks5roBoegaJpZM4Mc5w6>
> .
>

geoalgo commented:
I can add the test in another pr
